filament/backend/src/vulkan/VulkanSwapChain.h

```diff
@@ -14,81 +14,80 @@
-	    VkExtent2D fallbackExtent = {.width = 640, .height = 320});
-
-    // Headless constructor.
-    VulkanSwapChain(VkDevice device, VkPhysicalDevice physicalDevice,
-	    uint32_t graphicsQueueFamilyIndex, VkQueue graphicsQueue, VmaAllocator allocator,
-	    std::shared_ptr<VulkanCommands> commands, VulkanContext const& context,
-	    VulkanStagePool& stagePool, uint32_t width, uint32_t height);
-
-    bool acquire();
-    void destroy();
-    void recreate();
-    void makePresentable();
-    bool hasResized() const;
-
-    VulkanTexture& getColorTexture();
-    VulkanTexture& getDepthTexture();
-    uint32_t getSwapIndex() const { return mCurrentSwapIndex; }
-
-    VkSurfaceKHR surface = {};
-    VkSwapchainKHR swapchain = {};
-    VkSurfaceFormatKHR surfaceFormat = {};
-    VkExtent2D clientSize = {};
-    VkQueue presentQueue = {};
-    VkQueue headlessQueue = {};
-
-    // This is signaled when vkAcquireNextImageKHR succeeds, and is waited on by the first
-    // submission.
-    VkSemaphore imageAvailable = VK_NULL_HANDLE;
-
-    // This is true after the swap chain image has been acquired, but before it has been presented.
-    bool acquired = false;
-
-    bool suboptimal = false;
-    bool firstRenderPass = false;
+    VulkanSwapChain(VulkanPlatform* platform, VulkanContext const& context, VmaAllocator allocator,
+            std::shared_ptr<VulkanCommands> commands, VulkanStagePool& stagePool,
+            void* nativeWindow, uint64_t flags, VkExtent2D extent = {0, 0});
+
+    ~VulkanSwapChain();
+
+    void present();
+
+    void acquire(bool& reized);
+
+    inline std::shared_ptr<VulkanTexture> getCurrentColor() const noexcept {
+        return mColors[mCurrentSwapIndex];
+    }
+
+    inline std::shared_ptr<VulkanTexture> getDepth() const noexcept {
+        return mDepth;
+    }
+
+    inline bool isFirstRenderPass() const noexcept {
+        return mIsFirstRenderPass;
+    }
+
+    inline void markFirstRenderPass() noexcept {
+        mIsFirstRenderPass = false;
+    }
+
+    inline VkExtent2D getExtent() noexcept {
+        return mExtent;
+    }
 
 private:
-    void create();
+    void update();
 
-    VkDevice mDevice;
-    VkPhysicalDevice mPhysicalDevice;
+    VulkanPlatform* mPlatform;
     std::shared_ptr<VulkanCommands> mCommands;
     VmaAllocator mAllocator;
-    VulkanContext const& mContext;
     VulkanStagePool& mStagePool;
-    uint32_t mCurrentSwapIndex = 0u;
-    const VkExtent2D mFallbackExtent = {};
-
-    // Color attachments are swapped, but depth is not. Typically there are 2 or 3 color attachments
-    // in a swap chain.
-    utils::FixedCapacityVector<std::unique_ptr<VulkanTexture>> mColor;
-    std::unique_ptr<VulkanTexture> mDepth;
+    bool const mHeadless;
+
+    // We create VulkanTextures based on VkImages. VulkanTexture has facilities for doing layout
+    // transitions, which are useful here. We use std::shared_ptr because they will be shared with
+    // VulkanRenderTarget.
```

Comment:
Do we really need the thread-safety of shared_ptr? (i.e. that the counters are atomics)? that can get pretty expensive.

Reply:
changed TODO comment to reflect that smart pointer needs to be worked over in the vulkan driver.  Also filed #6838 6838